c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1470,290 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+  }
+};
+// The Coroutines spec has no `_::FixVoid<T>` equivalent to unify valueful and valueless co_return
+// statements, and programs are ill-formed if the coroutine implementation object (Coroutine<T>) has
+// both a `return_value()` and `return_void()`. No amount of EnableIffery can get around it, so
+// these return_* functions live in a CRTP mixin.
+
+template <typename T>
+class Coroutine<T>::Adapter: private Event {
+public:
+  explicit Adapter(PromiseFulfiller<T>& f, Coroutine::Handle coroutine)
+      : fulfiller(f), coroutine(coroutine) {
+    coroutine.promise().adapter = this;
+  }
+  KJ_DISALLOW_COPY(Adapter);
+
+  ~Adapter() noexcept(false) {
+    // If a coroutine runs to completion, its frame will already have been destroyed by the time the
+    // adapter is destroyed, so this if-condition will almost always be false.
+    if (coroutine) {
+      // coroutine.promise() is a reference to our implementation object, Coroutine<T>.
+      coroutine.promise().adapter = nullptr;
+      coroutine.destroy();
+    }
+  }
+
+  void coroutineDestroyed() {
+    // The coroutine was destroyed before the adapter.
+    coroutine = nullptr;
+  }
+
+  void awaitSuspend(PromiseNode& nodeParam, ExceptionOrValue& resultParam) {
+    // Called by Awaiter::await_suspend() to tell us to wait for `node` to become ready, then store
+    // the node's result in `result`.
+
+    // We'll need to save these references for fire(). Note that Awaiter is movable, and `result` is
+    // its subobject, but it is safe to store a reference to `result` in Adapter, because by the
+    // time `await_suspend()` is called, Awaiter has already been returned from `await_transform()`
+    // and stored in its final resting place in the coroutine's frame.
+    node = &nodeParam;
+    result = &resultParam;
+
+    // We use the promise adapter as an Event, because we must be able to destroy the coroutine from
+    // within the fire() callback. The promise adapter can do this, because it lives outside of the
+    // coroutine's frame, unlike Coroutine<T> or Awaiter<U>.
+    node->onReady(this);
+  }
+
+  PromiseFulfiller<T>& fulfiller;
+  // Used to fulfill/reject the enclosing coroutine's kj::Promise<T>.
+
+private:
+  Maybe<Own<Event>> fire() override {
+    // The promise this coroutine is currently waiting on is ready. Extract its result and check for
+    // exceptions.
+
+    // Destroying or resuming the coroutine will leave these pointers dangling, so we call this
+    // function before either action.
+    auto nullifyPointers = [&]() {
+      node = nullptr;
+      result = nullptr;
+    };
+
+    node->get(*result);
+
+    // Note: the Promise::wait() implementation throws a recoverable exception if both a value and
+    // an exception are set. We only have visibility on the exception here, so the following might
+    // not have quite the same semantics.
+
+    // Check for exceptions. We could let Awaiter::await_resume() do it, but then we'd resume the
```

Comment:
What if the coroutine contains a try/catch wrapping this await?

Reply:
Actually yeah, I think handling exceptions here is explicitly wrong. You always want to resume, and then in the exception case `await_resume()` should throw the exception, so that it can potentially be caught by a try/catch?

Reply:
Aw, crap.

Reply:
Admittedly, if we could somehow know that the coroutine _doesn't_ contain a try/catch -- which we'd expect to be the case the vast majority of the time -- then what you have here is probably a nice optimization, given how atrociously slow C++ exception unwinding is. I'm almost inclined to prohibit try/catch in our coroutines for this reason.

Though, then again, it's very possible that a premature `destroy()` is literally implemented using the same unwind mechanism as exceptions, in which case, maybe it doesn't make so much difference.

Reply:
I was thinking about this, and with -fno-exceptions, this incorrect optimization would be the required code path. But yeah, I think for the common case I'll need to move exception handling to throw from await_resume().

Reply:
Fixed: https://github.com/capnproto/capnproto/pull/1104/commits/5a9b13e4f9cd3312ad5f4ddad7c5f894d4ac5797